SPI: Add API support for DDR, Dual, Quad and Octal modes

tbursztyka opened the pull request:
(#20564 could not be reopened, so creating a new one.)

SPI extended modes
================

DDR: Double Data Rate
Dual: Dual MOSI lines
Quad: Quad MOSI lines
Octal: Octal MOSI lines
DQO: Dual Quand Octal

These modes comprise: DDR, Dual, Quad, Octal and the
specific command/address/data format


Uses cases
---------------

- single MOSI line: cmd / addr / data ddr
- single MOSI line: cmd / addr ddr / data ddr
- single MOSI line: cmd ddr / addr ddr / data ddr
- single MOSI line: cmd ddr
- single MOSI line: data ddr
- DQO MOSI lines: cmd single/addr single/data
- DQO MOSI lines: cmd single/addr/data
- DQO MOSI lines: cmd/addr/data
- DQO MOSI lines: cmd
- DQO MOSI lines: data
- DBO uses cases with DDR


Design
---------

Besides the operation attribute of struct spi_config, all goes through
a newly added u16_t attribute in struct spi_buf: flags.

Firts 9 bits are containing config flags, next 2 bits are the command
length (set to 0 if the cmd/addr format is not in use) and the last 5 bits
are the address length (set to 0 if the addr is not present or if the cmd/addr
format is not in use). Command length is a multiple of 4 (4, 8, 12, 16, ...)
and address length is a multiple of 2 factor 4 (4, 8, 16 ...)

**The good side of this approach:** it does not require changes on any existing SPI controller
drivers nor SPI device drivers.

**The drawback:** if only one SPI controller can support these modes, it adds up
16bits to each and every struct spi_buf in the system. Hopefully this is still
a relatively small amount of data (looks like uncommon for a driver to use more
than 2 struct spi_buf, and if does it will usually be due to a lack of factorization).

SPI configuration
----------------------

```
spi_config {
	   ...
	   .operation = SPI_LINES_<SINGLE/DUAL/QUAD/OCTAL> | SPI_EXTENDED_MODES,
	   ...
}
```

Basic cmd/addr format transaction usage
----------------------------------------------------

Note: 8bits length cmd/addr are used for these examples, but it's trivial to
use other size.

**TX only:**

```
u8_t tx_buf[] = { CMD, ADDR };

struct spi_buf buf[2] =  {
	{
		.buf = tx_buf,
       		.len = 2,
		.flags = SPI_EM_CMD_<mode> | SPI_EM_ADDR_<mode> |
       	      	       SPI_EM_CMD_LENGTH_SET(SPI_EM_CMD_LEN_8_BITS) |
		       SPI_EM_ADR_LENGTH_SET(SPI_EM_ADR_LEN_8_BITS)
	},
	{
		.buf = data,
 		.len = data_length
	}
};

struct spi_buf_set tx {
       .buffers = buf,
       .count = 2
};

spi_transceive(spi_dev, spi_cfg, &tx, NULL)
```

**RX added**

```
u8_t tx_buf[] = { CMD, ADDR };

struct spi_buf buf_tx[1] =  {
	{
		.buf = tx_buf,
       		.len = 2,
		.flags = SPI_EM_CMD_<mode> | SPI_EM_ADDR_<mode> |
       	      	       SPI_EM_CMD_LENGTH_SET(SPI_EM_CMD_LEN_8_BITS) |
		       SPI_EM_ADR_LENGTH_SET(SPI_EM_ADR_LEN_8_BITS)
	}
};

struct spi_buf_set tx {
       .buffers = buf_tx,
       .count = 1
};

struct spi_buf buf_rx[1] =  {
	{
		.buf = data,
 		.len = data_length
	}
};

struct spi_buf_set rx {
       .buffers = buf_rx,
       .count = 1
};

spi_transceive(spi_dev, spi_cfg, &tx, &rx)
```

**In case no address is given**

```
u8_t tx_buf[] = { CMD };

struct spi_buf buf_tx[1] =  {
	{
		.buf = tx_buf,
       		.len = 2,
		.flags = SPI_EM_CMD_<mode> | SPI_EM_ADDR_<mode> |
       	      	       SPI_EM_CMD_LENGTH_SET(SPI_EM_CMD_LEN_8_BITS) |
		       SPI_EM_ADR_LENGTH_SET(SPI_EM_ADR_NONE)
	}
};
```

etc ...

Commit:
spi: Add API support for DDR, Dual, Quad and Octal modes

These modes are present on more advanced and recent SPI controller.
Though dual, quand and octal were present on configuration side already,
it could not work with the standard command/address/data format which is
now addressed via these changes.

Also adding a way to make use of hardware optimizations for SPI flash
JEDEC devices which seem to be largely present in various SPI
controllers.

Note, however, that these new SPI API modes are not meant to be used by
QSPI flash controllers: for these, a dedicated flash controller driver
will still be the preferred solution.

Fixes #17902

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

carlescufi commented:
@tbursztyka please rebase

MaureenHelm commented on include/drivers/spi.h:
```diff
@@ -376,17 +376,120 @@ struct spi_dt_spec {
 #endif
 
+/**
+ * SPI Enhanced Mode flags: Double Data Rate, Dual/Quand/Octal modes
```
s/Quand/Quad/

MaureenHelm reviewed (dismissed):
As we discussed in the dev-review meeting, would like to see some documentation that explains the difference between this and dedicated QSPI/OSPI flash controllers

tbursztyka commented:
> As we discussed in the dev-review meeting, would like to see some documentation that explains the difference between this and dedicated QSPI/OSPI flash controllers

Sure will do. In fact, there are SPI controllers that do support d/q/o modes along with the usual ones. So they can (and they happen to be) wired with single line device (sensors, adapters etc...) and also a flash device. I told about it in API meeting but indeed it needs to be written directly to avoid confusion: native Q/O SPI flash controllers will need a dedicated flash driver as it is now. This will not replace it.

henrikbrixandersen reviewed (commented):
> **The good side of this approach:** it does not require changes on any existing SPI controller
drivers nor SPI device drivers.

Wouldn't non-static `struct spi_buf` instances contain unitialized `.flags`?

tbursztyka commented:
> > **The good side of this approach:** it does not require changes on any existing SPI controller
> > drivers nor SPI device drivers.
> 
> Wouldn't non-static `struct spi_buf` instances contain unitialized `.flags`?

I meant that if the controller does not support these new modes: it will not require any modification.

Now for the API users, uninitialized flags for operations that do not involve dual/quad/octal mode: it should not be an issue since this attribute will be then useless. Obviously, on dual/quad/octal mode case: user will need to set this attribute, at least to 0 (perhaps this needs to be documented). Or do you see something that could be a problem?

henrikbrixandersen commented:
> I meant that if the controller does not support these new modes: it will not require any modification.

Right.

> Now for the API users, uninitialized flags for operations that do not involve dual/quad/octal mode: it should not be an issue since this attribute will be then useless. Obviously, on dual/quad/octal mode case: user will need to set this attribute, at least to 0 (perhaps this needs to be documented). Or do you see something that could be a problem?

What if the API user does not require dual/octal/quad mode (and doesn't not set flags since it is existing client code, causing the flags to be unitialized), but the controller _does_ support a number of these flags. Now the client is suddenly passing flags to the controller that it didn't intend to pass?

glaubermaroto commented on include/drivers/spi.h:
```diff
@@ -376,17 +379,120 @@ struct spi_dt_spec {
+#define SPI_EM_CMD_DDR		BIT(2)
+/**
+ * SPI Addressing phase is send via standard mode (single line)
+ * spi_config's lines bit field will be superceded by this flag only
```
superseded

tbursztyka commented:
> What if the API user does not require dual/octal/quad mode (and doesn't not set flags since it is existing client code, causing the flags to be unitialized), but the controller _does_ support a number of these flags. Now the client is suddenly passing flags to the controller that it didn't intend to pass?

The driver must to ignore those. There is no reason for it to look at those flags if it's not configured in dual/quad/octal mode.

anangl commented on include/drivers/spi.h:
```diff
@@ -376,17 +379,120 @@ struct spi_dt_spec {
+ */
+#define SPI_EM_CMD_DDR		BIT(2)
+/**
+ * SPI Addressing phase is send via standard mode (single line)
+ * spi_config's lines bit field will be superseded by this flag only
+ * for the address.
+ */
+#define SPI_EM_ADR_STD		BIT(3)
+/**
+ * SPI Addressing phase is send via enhanced modes (double/quad/octal)
+ * The mode will depend on the spi_config's lines bit field.
+ */
+#define SPI_EM_ADR_DQO		BIT(4)
```
If I understand it correctly, these two flags are mutually exclusive. If so, why not use instead a 2-bit wide bit field that could take the values: `*_NONE` (no addressing phase) , `*_STD`, and `*_DQO`? Perhaps it would be good to also merge the bit field with the DDR flag (and use `*_STD_DDR` and `*_DQO_DDR` values), as the DDR flag alone has no point when there is no addressing phase.
Then the address length bit field could be 4-bit long (as the `*_NONE` value would be no longer needed there) and the documentation could specify that this field is ignored if there is no addressing phase.
Same scheme could be used for specification of the command phase (it could be then made possible to configure the command length to 12 bits; I don't know if such option is useful, though).

tbursztyka replied:
To summarize:

- no addressing phase would be: !SPI_EM_ADR_STD && !SPI_EM_ADR_DQO 
- thus reducing address length to 4 bits (and default value would be then SPI_EM_ADR_LEN_4_BITS )
- instead of using SPI_EM_ADR_DDR directly, let's have dedicated SPI_EM_ADR_STD_DDR = (SPI_EM_ADR_STD | SPI_EM_ADR_DDR) and SPI_EM_ADR_DQO_DDR = (SPI_EM_ADR_DQO | SPI_EM_ADR_DDR) 
- same for command.

Makes sense

tbursztyka replied:
While revising the API according to your comments, I found out I forgot one important config : the wait cycles (or dummies) after a transmission and right before a reception. This is mandatory for letting the slave to reconfigure itself properly.

And that would take more than the 2 left over bits I have. Looks like I'll need to raise flags attribute to 32bits. Note that this is a non issue: the compiler aligns the struct to 4bytes, so the size isn't going to change whether I use uint16_t or uint32_t for flags attribute. in the end spi_buf structure is going to get bloated by 4 bytes anyway.

tbursztyka replied:
> If so, why not use instead a 2-bit wide bit field that could take the values: *_NONE (no addressing phase) , *_STD, and *_DQO? Perhaps it would be good to also merge the bit field with the DDR flag (and use *_STD_DDR and *_DQO_DDR values), as the DDR flag alone has no point when there is no addressing phase.

Actually, that's not possible: you still have the case where cmd and/or address are sent in enhanced modes (dual/quad/octal). Completely forgot about it: these are the defaults. So we need a way to tell if there is a cmd and/or address and the only way is through the length.

carlescufi commented:
> > > **The good side of this approach:** it does not require changes on any existing SPI controller
> > > drivers nor SPI device drivers.
> > 
> > 
> > Wouldn't non-static `struct spi_buf` instances contain unitialized `.flags`?
> 
> I meant that if the controller does not support these new modes: it will not require any modification.
> 
> Now for the API users, uninitialized flags for operations that do not involve dual/quad/octal mode: it should not be an issue since this attribute will be then useless. Obviously, on dual/quad/octal mode case: user will need to set this attribute, at least to 0 (perhaps this needs to be documented). Or do you see something that could be a problem?

API Meeting:

- The issue is not the current set of flags introduced in this PR, but the fact that we now have a field, `flags` that, in the future, might contain flags that are used regardless of Kconfig options, and  in its current state this PR will leave SPI users that do not initialize this new `flags` field at all.

tbursztyka commented on include/drivers/spi.h:
```diff
@@ -376,17 +379,120 @@ struct spi_dt_spec {
+	(((_cfg_) & SPI_EM_ADR_LENGTH_MASK) >> SPI_EM_ADR_LENGTH_SHIFT)
+
+#define SPI_EM_ADR_LENGTH_SET(_len_)		\
+	((_len_) & SPI_EM_ADR_LENGTH_MASK)
```
indeed

tbursztyka commented:
@carlescufi @henrikbrixandersen ah actually you were right, there is an uninitialized flags issue. On single line (default), controller driver will not know if DDR, command/addressing phases are required: and as soon as CONFIG_SPI_ENHANCED_MODES is enabled it will thus need to look there. That was eluding me, good you asked multiple times.

So to prevent that, I'll add another config bit (in the extended  32bits operation attribute, there are left overs): 0 (by default) will mean no enhanced modes. So all existing SPI device drivers will not be affected by the flags attribute being added. 1 will enable the enhanced modes, thus the controller driver will look at flags if that bit is set.

tbursztyka commented:
Updated the API. 
- I remove the JEDEC specific flag, that a left over when this proposal was competing with qspi dedicated driver. That's no longer necessary.
- added the SPI_EXTENDED_MODES configuration bit, which is unset by default. Preventing uninitialized spi_buf flags issue. (SPI controller driver will not look at the flags attribute if such config bit is unset).

I'll try to find some time to update DW driver and flash's spi_nor.c  (intel s1000 could take advantage of this: its generic SPI controller, which can provide dual/quad modes, can be - it is on the board I have - wired to SPI flash allowing dual/quad), to show how this API could be used.

henrikbrixandersen commented:
Looks good. It would be nice to have an in-tree implementation added as well to get some testing of this API extension.

carlescufi commented:
@tbursztyka can you rebase please?

carlescufi commented:
@tbursztyka  please rebase

tbursztyka commented:
@scottwcpg I am late in answering your comment

> A question/comment. Current Zephyr SPI is based on full-duplex. If you look at the flash spi_nor.c driver, it provides an RX buffer for each TX buffer. Dual/Quad/Octal are inherently half-duplex. To me, I do not know how I would implement Dual/Quad in an SPI driver if I had to duplicate full duplex behavior in how the SPI buffers are handled. If all I/O lines are set to output during transmit it makes no sense to fill a corresponding receive buffer with what? User's and SPI driver developers need clear documentation in the SPI header and SPI API on how to setup SPI buffers for full-duplex, half-duplex(3-pin), and Dual/QuadOctal. Thanks.

If you are only transmitting, you will not need to provide any rx buffer. Thus, actually, the spi_read()/spi_write() helpers: they just call transceive with, respectively, a NULL tx buffer or a NULL rx buffer.

cfriedt commented:
@tbursztyka - this is super useful. To be honest, I'm surprised it was not added a long time ago. I would be happy to review again when test cases are passing.

albertofloyd commented:
FYI @rgujju

tbursztyka commented:
> @tbursztyka - this is super useful. To be honest, I'm surprised it was not added a long time ago. I would be happy to review again when test cases are passing.

No strong use-case until now I guess.
Also, I failed to find time to enable this feature on dw driver (for instance) to show how well/bad it works. So this will require a thorough review

carlescufi commented:
@erwango @teburd @galak please take another look

teburd commented:
I did a brief comparison with the NXP FlexSPI peripheral to get a sense of how closely matching this is.

There's similarities there but some worthwhile differences to this API.

In effect FlexSPI works through a command queue separate from queue tx/rx buffers. The command queue has 16 bit operations that consist of

6 bit op code, 2 bit pad count, 8 bit operand.

This setup mostly matches up with the EM flags proposed here though... in the FlexSPI case each command decides the number of lines to use, so single, dual, quad, or octal is decided by the 2 bit pad count. I don't see the equivalent in the flag set here. Instead there's a SPI_EM_DQO flag which I'd assume means its decided at compile time somehow (via device tree per node or via Kconfig) for the entire set of commands or device on the bus in some manner.

There's also the row and column address commands which take an 8 bit operand for the address length rather than a fixed set of address lengths as is done here, I think this might be the most significant difference I can spot.

The op codes themselves are maybe interesting to note and there are many more than available than the EM flags here that might be interesting to consider.

I'll try and peruse a few more SPI IP blocks and compare and contrast as well this week.

tbursztyka commented:
@teburd  interesting, I wrongly assumed that d/q/o modes were somewhat standardized across manufacturers and thus took only one hardware spec as a reference. I fail to see any use case that would require changing modes within the transaction but better be prepared (after all, devices using these modes - spi flash aside - are yet to become a common thing).

scottwcpg commented:
I did not see a flag for data phase number of pin. Is the data phase using the same number of pins as the address phase?
Also, should we discuss a common approach for SPI flash "dummy" clocks and mode bytes. Most SPI flash devices recommend the only the first 3 or 4 bits of the mode byte are driven as data, the remaining bits are tri-state. Byte oriented SPI controllers can only drive a whole byte so a split mode byte can be done.  Our QMSPI controller has a bit mode that can split a mode byte. Is it worth the effort for one controller?  Probably not.

teburd commented:
I looked briefly at the SAM QSPI and it seems the flags match better there.

For SAM QSPI a seperate QSPI_IAR (Instruction Address Register) is available which can be used to setup an optional instruction frame before read/writing data to the QSPI lines.

It offers optional fields that may be included for an instruction (8 bits), address (24 or 32 bit), and option (4-8 bits) written using 1 to 4 datalines each followed by reading/writing data in single/dual/quad spi.

I'm not entirely sure this supports all the flags but I think at least some subset of the flags here would work.

tbursztyka commented:
> I'm not entirely sure this supports all the flags but I think at least some subset of the flags here would work.

As per-controller, it does have to support all the flags, it's fine. My concern is that the API should enable (most of) all the controllers (to a certain generalization extent).

> I did not see a flag for data phase number of pin. Is the data phase using the same number of pins as the address phase? 

There are flags for this. See SPI_EM_CMD_STD or SPI_EM_ADDR_STD for instance. It is either one line, or the whole set of lines.
I am not sure there is a need for more flexibility (2 lines for address, and 4 for data for instance). I lacked of verifiable use-cases when designing this API.

>Also, should we discuss a common approach for SPI flash "dummy" clocks and mode bytes. Most SPI flash devices recommend the only the first 3 or 4 bits of the mode byte are driven as data, the remaining bits are tri-state. Byte oriented SPI controllers can only drive a whole byte so a split mode byte can be done. Our QMSPI controller has a bit mode that can split a mode byte. Is it worth the effort for one controller? Probably not.

The effort with this PR is to provide dual/quad/octal API exposed by generic SPI controllers, i.e. ones that can be wired to any type of SPI devices. Flash included BUT: I originally added flags for flash specific optimizations but it was not as good as doing a qspi flash driver from performance/flexibility point of view. I am so sure it's worth the generalization effort to include flash idiosyncrasies, moreover that SPI controllers offers optimizations at different level compared to others. We could perhaps expose something as useful as XIP, but fine tuning options for flash... not sure.

MaureenHelm commented:
@tbursztyka can you implement this in a driver and demonstrate its use in an application?

tbursztyka commented:
> @tbursztyka can you implement this in a driver and demonstrate its use in an application?

Sure, I would need to find some time for it. But first, I need to find a platform that has a generic SPI controller capable of supporting dual/quad/octal as well, and a device to make transactions with.

MaureenHelm commented:
> > @tbursztyka can you implement this in a driver and demonstrate its use in an application?
> 
> Sure, I would need to find some time for it. But first, I need to find a platform that has a generic SPI controller capable of supporting dual/quad/octal as well, and a device to make transactions with.

Can you move this into the draft state until then?

cfriedt commented:
> Sure, I would need to find some time for it. But first, I need to find a platform that has a generic SPI controller capable of supporting dual/quad/octal as well, and a device to make transactions with.

@tbursztyka - this is a really interesting topic. I wonder if it might even be possible to list some relatively simple SPI peripherals that support each mode, and then maybe we could look into `spi_emul` and peripheral emulation for the time being.

cfriedt commented:
dev-review: @tbursztyka - do you have plans to move forward with this?

MaureenHelm commented:
> dev-review: @tbursztyka - do you have plans to move forward with this?

@tbursztyka ping

tbursztyka commented:
Reopening for further investigation related to #65659
